src/prefect/engine/cloud/flow_runner.py

```diff
@@ -118,8 +119,25 @@ def call_runner_target_handlers(self, old_state: State, new_state: State) -> Sta
         try:
             cloud_state = new_state
             state = self.client.set_flow_run_state(
-                flow_run_id=flow_run_id, version=version, state=cloud_state
+                flow_run_id=flow_run_id,
+                version=version if cloud_state.is_running() else None,
+                state=cloud_state,
             )
+        except VersionLockError:
+            state = self.client.get_flow_run_state(flow_run_id=flow_run_id)
+
+            if state.is_running():
+                self.logger.debug(
+                    "Version lock encountered and flow is already in a running state."
+                )
+                raise ENDRUN(state=state)
+
+            self.logger.debug(
+                "Version lock encountered, proceeding with state {}...".format(
+                    type(state).__name__
+                )
+            )
+            new_state = state
```

Comment:
Should this `ENDRUN` as well, to abort further processing of this task? For example, let's say it tries to enter a `RUNNING` state, but learns that the "true" state is Failed. I don't think we want to enter the `RETRYING` part of this task runner's pipeline, which we might if we just return the upstream state. Therefore I'm wondering if anytime we learn that the true state is different from what we thought, we just ENDRUN with it and let the next task proceed. 

Not 100% sure if I'm thinking through everything here - what do you think?